protocol/protocol.go

```diff
@@ -0,0 +1,131 @@
+package protocol
+
+import (
+	"context"
+	"fmt"
+
+	logging "github.com/ipfs/go-log"
+	"github.com/jbenet/goprocess"
+	goprocessctx "github.com/jbenet/goprocess/context"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/libp2p/go-smart-record/ir"
+	pb "github.com/libp2p/go-smart-record/protocol/pb"
+	"github.com/libp2p/go-smart-record/vm"
+	sm "github.com/libp2p/go-smart-record/vm"
+)
+
+var log = logging.Logger("smart-records")
+
+const (
+	srProtocol protocol.ID = "/smart-record/0.0.1"
+)
+
+// SmartRecordManager interface to manage smart records
+type SmartRecordManager interface {
+	Get(ctx context.Context, k string, p peer.ID) (*vm.Record, error)
+	Update(ctx context.Context, k string, p peer.ID, rec ir.Dict) error
+	// NOTE: we won't support queries until we figure out selectors
+	// Query(ctx context.Context, k string, p peer.ID, selector ir.Dict) (*ir.Dict, error)
+}
+
+// smartRecordManager handles the exchange of messages
+// with peers to interact with smart-records
+type smartRecordManager struct {
+	ctx       context.Context
+	proc      goprocess.Process
+	host      host.Host
+	self      peer.ID
+	vm        vm.Machine
+	protocols []protocol.ID
+
+	senderManager *messageSenderImpl
+}
+
+func NewSmartRecordManager(ctx context.Context, h host.Host, options ...Option) (SmartRecordManager, error) {
+	return newSmartRecordManager(ctx, h, options...)
+}
+
+func newSmartRecordManager(ctx context.Context, h host.Host, options ...Option) (*smartRecordManager, error) {
+	var cfg config
+	if err := cfg.apply(append([]Option{defaults}, options...)...); err != nil {
+		return nil, err
+	}
+	protocols := []protocol.ID{srProtocol}
+
+	e := &smartRecordManager{
+		ctx:       ctx,
+		proc:      goprocessctx.WithContext(ctx),
+		host:      h,
+		self:      h.ID(),
+		vm:        sm.NewVM(cfg.mergeContext, cfg.assembler),
```

Comment:
Field vm is never used.

Reply:
It is used in `protocol/handler.go` to interact with the VM when you receive requests from other peers.